rpc/lib/types/types.go

```diff
@@ -13,17 +13,63 @@ import (
 	tmpubsub "github.com/tendermint/tendermint/libs/pubsub"
 )
 
+// a wrapper to emulate a sum type: jsonrpcid = string | int
+// TODO: refactor when Go 2.0 arrives https://github.com/golang/go/issues/19412
+type jsonrpcid interface {
+	isJSONRPCID()
+}
+
+// JSONRPCStringID a wrapper for JSON-RPC string IDs
+type JSONRPCStringID string
+
+func (JSONRPCStringID) isJSONRPCID() {}
+
+// JSONRPCIntID a wrapper for JSON-RPC integer IDs
+type JSONRPCIntID int
+
+func (JSONRPCIntID) isJSONRPCID() {}
+
 //----------------------------------------
 // REQUEST
 
 type RPCRequest struct {
 	JSONRPC string          `json:"jsonrpc"`
-	ID      string          `json:"id"`
+	ID      jsonrpcid       `json:"id"`
```

Comment:
One interesting thing to note is that the `json` pkg can natively unmarshal into empty interfaces, but not into interfaces with methods. Crazy, I know, it's one of the main reasons we invented Amino.

In any case, I like the type strength here, but when we actually unmarshal it might be nicer to unmarshal into an anonymous struct like:

```
struct {
	JSONRPC string          `json:"jsonrpc"`
	ID      interface{}       `json:"id"`
	Method  string          `json:"method"`
	Params  json.RawMessage `json:"params"`
}
```

Then we can populate a proper RPCRequest with these fields, switching on the ID to see if we should use JSONRPCIntID or JSONRPCStringID.

I like this better than using the errors because we don't know if the error was from a real failure to unmarshal, or just from using a different ID type, and then we blindly just try again. So marshalling into the interface{} and checking the type feels more correct.

What do you think?

Reply:
I checked `Allow edits from maintainers`, so feel free to push to the `fix/rpc-integer-id` branch on my fork.

Reply:
From the Go style guides I've seen, they usually recommend avoiding empty interfaces, but since this is the first Go code I ever wrote, I don't have any strong opinion for or against that.

Reply:
changed here: https://github.com/tendermint/tendermint/pull/2811/commits/f2a71d7b9dd636816c0da35384779390ff1a3338#diff-f595cd0be03b00ec17ec43d65afb7780 -- needed nil check, since ID may not be provided (e.g. in ws_client_test)